hasktorch/src/Torch/Data/Pipeline.hs

```diff
@@ -0,0 +1,79 @@
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE PartialTypeSignatures #-}
+module Torch.Data.Pipeline where
+
+import Pipes
+import qualified Pipes.Prelude as P
+import Pipes.Concurrent
+import Control.Concurrent.Async
+import Control.Monad
+import Control.Arrow (first)
+
+
+type Iter = Int
+type WorkerId = Int
+
+data DatasetMock m tensor = DatasetMock { getBatchMock :: Int -> m tensor
+                                        , numIters :: Iter
+                                        }
+
+class ConcurrentDataset dataset numWorkers where
+  getBatchConcur :: Iter -> WorkerId -> IO tensor
+
+class Dataset dataset tensor where
+  getBatch :: Iter -> IO tensor
+
+data RunBatch = Final | KeepTrain  deriving (Eq, Show)
+
+takeBatch :: MonadIO m => Input (batch, RunBatch) -> Producer batch m ()  
+takeBatch input = fromInput input >-> P.takeWhile ((/=) Final . snd) >-> P.map fst
+
+readBatches ::
+  MonadIO m => DatasetMock m (tensor, RunBatch) -> Output (tensor, RunBatch)  -> Effect m () 
+readBatches DatasetMock{..} transformBox = do
+  -- TODO: might want to pair the RunBatch value in this function
+  for (each [1..numIters]) (\iter -> yieldBatch iter >-> toOutput transformBox)
+    where yieldBatch iter = (lift $ getBatchMock iter) >>= yield
+
+runTransforms :: MonadIO m => (tensor -> tensor') -> Input (tensor, RunBatch) -> Output (tensor', RunBatch) -> Effect m ()
+runTransforms transforms transformBox trainBox = fromInput transformBox >->  P.map (first transforms) >-> toOutput trainBox
+
+makeFoldWithTransform ::
+  _ => (batch -> batch')
+  -> DatasetMock IO (batch, RunBatch)
+  -> Int
+  -> IO ((b -> batch' -> m b) -> b -> m b) 
+makeFoldWithTransform transforms dataset numEpochs = createTrainLoop
+  where createTrainLoop  =  do
+          -- TODO: we can allow different buffer sizes
+          -- which would be necessary for data echoing
+            (toTransformBox, fromTransformBox, sealTransform) <- spawn' (bounded 1)
+            (toBatches, fromBatches, sealBatch) <- spawn' (bounded 1)
+            batchReader <-  async $ do runEffect $ replicateM_ numEpochs $ readBatches dataset toTransformBox 
+            transformer <-  async $ do runEffect $ runTransforms transforms fromTransformBox toBatches
+                                       atomically sealTransform
+                                       atomically sealBatch
```

Comment:
batchReader and transformer are not  used. Do we need to check the returned Async?

Reply:
This is a remnant from how I was doing things earlier. As it is now I don't
think there is a need to use the returned async values, since these threads
shouldn't terminate until the thread which is receiving data stops asking for data
(And even then, I think the threads would sit idle if the thread receiving batches asked to receive less epochs of data than are yielded here).

Reply:
Looking at this again has helped me realize something though. In the batch
reading thread we should be able to repeat the action infinitely (rather than
numEpochs number of times), and then force the clean up of those streams
after the fold has finished. This should make it so that we couldn't have a Consumer sitting waiting for data that would never be yielded, nor a Producer waiting to yield data forever.